rpch-crypto/src/crypto.rs

```diff
@@ -0,0 +1,463 @@
+    InvalidSession,
+    #[error("message verification failed")]
+    VerificationFailed,
+    #[error("private key is missing in the used identity")]
+    MissingIdentityKey,
+    #[error("error while parsing the request or response message")]
+    MessageParseError,
+    #[error("unsupported protocol version")]
+    UnsupportedProtocolVersion,
+    #[error("invalid counter value")]
+    InvalidCounter,
+    #[error("low level cryptographic error: {0}")]
+    CryptographicError(String),
+    #[error("not implemented")]
+    NotImplemented
+}
+
+type Result<T> = core::result::Result<T, RpchCryptoError>;
+
+pub struct Session {
+    req_data: Option<Box<[u8]>>,
+    resp_data: Option<Box<[u8]>>,
+    counter: u64,
+    shared_presecret: Option<SharedSecret>
+}
+
+impl Session {
+
+    pub fn get_request_data(&self) -> Option<Box<[u8]>> {
+        self.req_data.clone()
+    }
+
+    pub fn get_response_data(&self) -> Option<Box<[u8]>> {
+        self.resp_data.clone()
+    }
+
+    pub fn counter(&self) -> u64 {
+        self.counter
+    }
+}
+
+pub struct Identity {
+    pubkey: PublicKey,
+    counter: Option<u64>,
+    secret_key: Option<SecretKey>
+}
+
+impl Identity {
+
+    pub fn new(public_key: &[u8], counter: Option<u64>, private_key: Option<Box<[u8]>>) -> Result<Identity> {
+        let pk = PublicKey::from_sec1_bytes(public_key)
+            .map_err(|e| RpchCryptoError::CryptographicError(e.to_string()))?;
+
+        let sk = match private_key {
+            None => None,
+            Some(k) => Some(SecretKey::from_be_bytes(k.as_ref())
+                .map_err(|e| RpchCryptoError::CryptographicError(e.to_string()))?)
+        };
+
+        Ok(Identity {
+            pubkey: pk,
+            counter,
+            secret_key: sk
+        })
+    }
+
+    fn counter(&self) -> Option<u64> {
+        self.counter
+    }
+}
+
+pub struct Envelope {
+    message: Box<[u8]>,
+    entry_peer_id: String,
+    exit_peer_id: String
+}
+
+impl Envelope {
+
+    pub fn new(message: &[u8], entry_peer_id: &str, exit_peer_id: &str) -> Envelope {
+        Envelope {
+            message: message.into(),
+            entry_peer_id: entry_peer_id.into(),
+            exit_peer_id: exit_peer_id.into()
+        }
+    }
+
+    pub fn entry_peer_id(&self) -> &str {
+        self.entry_peer_id.as_str()
+    }
+
+    pub fn exit_peer_id(&self) -> &str {
+        self.exit_peer_id.as_str()
+    }
+
+    pub fn message(&self) -> &[u8] {
+        self.message.as_ref()
+    }
+
+}
+
+const REQUEST_TAG: &str = "req";
+
+fn initialize_cipher(shared_presecret: &SharedSecret, counter: u64, salt: &[u8], start_index: usize) -> Result<(ChaCha20Poly1305, Vec<u8>)> {
+    let kdf = shared_presecret.extract::<Blake2s256>(Some(salt));
+
+    let mut key = [0u8; CIPHER_KEYSIZE];
+    let mut ivm = [0u8; CIPHER_IVSIZE - COUNTER_SIZE];
+    for _ in 0..start_index+1 {
+        // Generate the encryption key using the KDF
+        kdf.expand(&[0u8; 1], &mut key)
```

Comment:
Why don't use `start_index` directly? Don't think the API is meant to be called iteratively

Reply:
There's no way in the Rust `hkdf` API to use the `start_index` directly. According to the docs paper:
````
The most common way to use HKDF is as follows: you provide the Initial Key Material (IKM) and an optional salt, then you expand it (perhaps multiple times) into some Output Key Material (OKM) bound to an “info” context string.
````